File: Part 3 - Classification/Section 14 - Logistic Regression/log_reg.R
#importing dataset
df = read.csv('Social_Network_Ads.csv')
df = df[3:5]

#Splitting data into training and test data
library(caTools)
set.seed(1234)
split = sample.split(df$Purchased, SplitRatio = 0.75)
training = subset(df, split == TRUE)
test = subset(df, split == FALSE)

#Feature scaling
training[1:2] = scale(training[1:2])
test[1:2] =  scale(test[1:2])

#Fitting Model
classifier = glm(formula = Purchased ~ .,
                 family = binomial,
                 data = training)

#Predicting the results
prob_pred = predict(classifier, type = 'response', newdata = test[-3])
y_pred = ifelse(prob_pred > 0.5,
                1,
                0)

#Generating confusion matrix
cm = table(test[,3],y_pred)

#Visualisation of the result
#install.packages('ElemStatLearn')
library(ElemStatLearn)
set = training
X1 = seq(min(set[,1]) - 1, max(set[,1]) + 1, 0.01)
X2 = seq(min(set[,2]) - 1, max(set[,1]) + 1, 0.01)
grid_set = expand.grid(X1,X2)
colnames(grid_set) = c('Age','EstimatedSalary')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1 ,0)
plot(set[,-3],
     main = 'Logistic Regression(Training Set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.',col = ifelse(y_grid == 1, 'springgreen','tomato'))
points(set[,-3], pch =21, bg = ifelse(set[,3] == 1, 'green4', 'red3'))

#Visualisation of the result (test set)
set = test
X1 = seq(min(test[,1]) - 1, max(test[,1] + 1), 0.01)
X2 = seq(min(test[,2]) - 1, max(test[,2] + 1), 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age','EstimatedSalary')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[,-3],
     main = 'Logistic Regression(Test Set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)),add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[,3] == 1, 'green4','red4'))

File: Part 3 - Classification/Section 14 - Logistic Regression/log_reg.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 20 12:38:59 2018

@author: Vinayak_S
"""
#importing library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#Reading dataset
df = pd.read_csv('Social_Network_Ads.csv')
X = df.iloc[:,2:4].values
y = df.iloc[:,-1].values

#Splitting dataset into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

#Feature scaling
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
X_train = sc_x.fit_transform(X_train)
X_test = sc_x.transform(X_test)


#Fitting logistic regression model to the training data
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)


#Predicting the test set results
y_pred = classifier.predict(X_test)

#Making the confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

#Visualisation of the training set
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:,0].min() - 1, stop = X_set[:,0].max() + 1, step = 0.01), 
                     np.arange(start = X_set[:,1].min() - 1, stop = X_set[:,1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(),X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.5, cmap = ListedColormap(('red','green')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i,j in enumerate(np.unique(y_set)):
    plt.scatter(x = X_set[y_set == j,0], y =X_set[y_set == j, 1], 
                c = ListedColormap(('red','green'))(i),label = j)
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.title('Logistic Regression (Training Set)')
plt.legend()
plt.show()

#Visualisation of the test set
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(X_set[:,0].min() - 1, X_set[:,0].max() + 1, 0.01),
                     np.arange(X_set[:,1].min() - 1 ,X_set[:,1].max() + 1, 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.5, cmap = ListedColormap(('red','green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x = X_set[y_set == j,0], y = X_set[y_set == j,1], c = ListedColormap(('red','green'))(i),
                label = j)
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.title('Logistic Regression(Test Set)')
plt.legend()
plt.show()
